src/interrupts/mod.rs

```diff
@@ -0,0 +1,577 @@
+    InterruptAlreadyInUse(InterruptRequest),
+}
+
+/// The `InterruptHandle` is used to access and configure an active interrupt.
+pub struct InterruptHandle<T> {
+    _data_type: PhantomData<T>,
+    irq: InterruptRequest,
+}
+
+impl<T> InterruptHandle<T> {
+    fn new(irq: InterruptRequest) -> Self {
+        InterruptHandle { 
+            irq: irq,
+            _data_type: PhantomData,
+        }
+    }
+}
+
+/// The `InterruptTable` guarantees safe and 'free of data races' use of interrupts.
+///
+/// To ensure that no data races can occur, it uses the Send and Sync concurrency concept from Rust.
+/// The `InterruptTable` can only be used in the `code(&mut InterruptTable)` function passed to the `scope` function,
+/// to ensure that `InterruptTable.drop()` is called.
+///
+/// # Examples
+/// ```
+/// interrupts::scope(nvic, |irq| { hprintln!("Default handler: {}", irq) },
+///     use stm32f7::interrupts::interrupt_request::InterruptRequest::Tim7;
+///     use interrupts::Priority::P1;
+///     |interrupt_table| {
+///
+///         let _ = interrupt_table.register(Tim7, P1, || {
+///         hprintln!("Interrupt handler for Tim7");
+///     });
+/// });
+/// ```
+pub struct InterruptTable<'a> {
+    _lifetime: PhantomData<&'a ()>,
+    nvic: &'static mut Nvic,
+    data: [Option<* mut ()>; 98],
+}
+
+impl<'a> Drop for InterruptTable<'a> {
+    fn drop(&mut self) {
+        let mut some_left = false;
+        unsafe {
+            DEFAULT_HANDLER = None;
+            for (i,isr) in ISRS.iter().enumerate() {
+                some_left = some_left || isr.is_some();
+                self.disable_interrupt(i as u8);
+            }
+        }
+        if some_left {
+            panic!("Disable interrupts first");
+        }
+    }
+}
+
+/// Creates a new scope, to guarantee that the `InterruptTable` constructor is called.
+///
+/// # Examples
+/// ```rust
+/// fn main(hw: board::Hardware) -> ! {
+///     // Extract hardware
+///     let board::Hardware {
+///         rcc,
+///         pwr,
+///         flash,
+///         nvic,
+///         ..
+///     } = hw;
+///    
+///     // Configure system clock
+///     system_clock::init(rcc, pwr, flash);
+///
+///     use stm32f7::interrupts::interrupt_request::InterruptRequest::Tim7;
+///     use interrupts::Priority::P1;
+///     // Open scope with interrupt support
+///     interrupts::scope(nvic, |irq| { hprintln!("Default handler: {}", irq) },
+///         |interrupt_table| {
+///             let _ = interrupt_table.register(Tim7, P1, || {
+///                 hprintln!("Interrupt handler for Tim7");
+///         });
+///     });
+///    loop{}
+/// }
+/// ```
+///
+/// # Panics
+/// Panics if an interrupt is enabled and is not disabled after use in `code()`
+pub fn scope<'a,F,C,R>(nvic: &'static mut Nvic, default_handler: F, code: C) -> R 
+    where F: FnMut(u8) + 'a,
+        C: FnOnce(&mut InterruptTable<'a>) -> R
```

Comment:
Reading your thesis I wondered whether we can remove the `PhantomData` from `InterruptTable` and instead write all methods of `InterruptTable` like `fn foo<'a>(&'a mut self, ...)`. I think that would be simpler and convey the same semantics.